feat(dns): Add support for shared zones in Designate v2 API

victorpcamargo opened the pull request:
This commit introduces functionality to manage shared zones in the OpenStack Designate v2 API via Terraform.
It includes the following:

- New resource: openstack_dns_zone_share_v2 for creating, deleting and importing shared zones.
- New data source: openstack_dns_zone_share_v2 for listing shared zones by zone ID and target project.
- Updated provider.go to register the new resource and data source.
- New documentation about the new functionality.
- Comprehensive unit tests for both resource and data source.
- Examples for resource and data source usage.

These additions align the Terraform OpenStack provider with Designate’s shared zones feature, enabling users to share DNS zones across projects programmatically.

It closes issue https://github.com/terraform-provider-openstack/terraform-provider-openstack/issues/1692

kayrus commented:
@victorpcamargo I haven't read code yet, but I have a quick question, is it possible to implement a shared zone within an existing zone resource?

victorpcamargo commented:
> @victorpcamargo I haven't read code yet, but I have a quick question, is it possible to implement a shared zone within an existing zone resource?
> 
> UPD: got it. it's kinda zone transfer. Considering this I'd rename the resource from `openstack_dns_shared_zone_v2` to `openstack_dns_zone_share_v2`

Thanks @kayrus. That makes sense indeed. Code updated.

kayrus commented on openstack/resource_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,234 @@
+
+	// Log API details for debugging
+	fmt.Printf("DEBUG: Making API call to share DNS zone\n")
+	fmt.Printf("DEBUG: URL: %s\n", client.ServiceURL("zones", zoneID, "shares"))
+	fmt.Printf("DEBUG: Headers: X-Auth-Sudo-Project-Id: %s\n", sudoProjectID)
+	fmt.Printf("DEBUG: Request Body: %+v\n", shareOpts)
```
this is not necessary, if you enable debug logs, this info will be printed anyway.

kayrus commented on openstack/resource_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,234 @@
+// resourceDNSZoneShareV2 defines the resource to share a DNS zone.
+func resourceDNSZoneShareV2() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceDNSZoneShareV2Create,
+		Read:   resourceDNSZoneShareV2Read,
+		Delete: resourceDNSZoneShareV2Delete,
```
```suggestion
		CreateContext: resourceDNSZoneShareV2Create,
		ReadContext:   resourceDNSZoneShareV2Read,
		DeleteContext: resourceDNSZoneShareV2Delete,
```

victorpcamargo replied:
Changed to all of them.

kayrus commented on openstack/data_source_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,87 @@
+
+func dataSourceDNSZoneShareV2() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceDNSZoneShareV2Read,
```
```suggestion
		ReadContext: dataSourceDNSZoneShareV2Read,
```

kayrus commented on openstack/data_source_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,87 @@
+
+func dataSourceDNSZoneShareV2Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	client, err := config.DNSV2Client(context.Background(), "")
```
```suggestion
	client, err := config.DNSV2Client(ctx, "")
```

kayrus commented on openstack/resource_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,234 @@
+	return parts[0], parts[1], nil
+}
+
+func listZoneShares(client *gophercloud.ServiceClient, zoneID string, ownerProjectID string) ([]ZoneShare, error) {
```
```suggestion
func listZoneShares(ctx context.Context, client *gophercloud.ServiceClient, zoneID string, ownerProjectID string) ([]ZoneShare, error) {
```

kayrus commented on openstack/resource_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,234 @@
+		},
+	}
+
+	_, err := client.Get(context.Background(), url, &result, reqOpts)
```
```suggestion
	_, err := client.Get(ctx, url, &result, reqOpts)
```

kayrus commented on openstack/resource_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,234 @@
+	}
+
+	// Perform the delete request
+	_, err = client.Delete(context.Background(), url, reqOpts)
```
```suggestion
	_, err = client.Delete(ctx, url, reqOpts)
```

kayrus commented on openstack/resource_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,234 @@
+
+func resourceDNSZoneShareV2Delete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	client, err := config.DNSV2Client(context.Background(), "")
```
```suggestion
	client, err := config.DNSV2Client(ctx, "")
```

kayrus commented on openstack/resource_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,234 @@
+
+func resourceDNSZoneShareV2Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
+	client, err := config.DNSV2Client(context.Background(), "")
```
```suggestion
	client, err := config.DNSV2Client(ctx, "")
```

kayrus commented on openstack/data_source_openstack_dns_zone_share_v2.go:
```diff
@@ -0,0 +1,87 @@
+		return fmt.Errorf("error retrieving shared zones for DNS zone %s: %s", zoneID, err)
+	}
+
+	// Ensure we are parsing the correct JSON field
+	var results []map[string]interface{}
+	for _, share := range shares {
+		results = append(results, map[string]interface{}{
+			"share_id":   share.ID,
+			"project_id": share.TargetProjectID,
+		})
+	}
```
Maybe its better to make a flatten func from this

```suggestion
	// Ensure we are parsing the correct JSON field
	results := make([]map[string]string, 0, len(shares))
	for _, share := range shares {
		results = append(results, map[string]interface{}{
			"share_id":   share.ID,
			"project_id": share.TargetProjectID,
		})
	}
```

victorpcamargo replied:
New flatten func created for this purpose.

kayrus commented on docs/resources/dns_zone_share_v2.md:
```diff
@@ -0,0 +1,67 @@
+---
```
can you add a doc for the data source as well?

victorpcamargo replied:
Thanks @kayrus. Added a doc for datasource as well.

kayrus reviewed (commented):
Hi. Can you use context-aware func types instead of the `context.Background`? See also a comment about the docs.

As for the import, the `:` separator is not used anywhere in other resources. Can you use just a slash?

And please fix the failing DNS functional tests.

victorpcamargo commented:
@kayrus, thanks for the previous review. I think everything has been resolved now. For the tests, I need to share the zone with another project, but it seems that the user running the tests doesn't have permission to create a project, so I changed the tests to get a list of projects and try to share the zone with one of them. I hope that the user used for the tests will be able to do this.

kayrus commented:
@victorpcamargo you can run tests under the privileged user, see `testAccPreCheckAdminOnly(t)`

victorpcamargo commented:
@kayrus Thanks for the advice. I have reverted the tests to the project creation method and I believe it will now pass.

kayrus commented:
@victorpcamargo please rebase on main and remove the go.mod updates, we're already on the recent commit

victorpcamargo commented:
Thanks @kayrus. Just rebased the branch and updated go.mod according to the main, but you'll also see some other updates in go.mod file related to other required packages.

kayrus commented:
@victorpcamargo if you make a proper rebase, not a merge, the go.mod dependencies must go away from this PR.

kayrus commented:
@victorpcamargo can you try to rebase to avoid any changes to go.mod?

victorpcamargo commented:
@kayrus It looks good now. Could you pls review it?

victorpcamargo commented:
Hi @kayrus. Am I missing something? Pls lmk if everything is ok and if we can marge it. Thanks!

kayrus commented:
@victorpcamargo apologies, I'm extremely busy this week

victorpcamargo commented:
Don't worry, @kayrus. I just wanted to know if I was missing anything, since I haven't heard anything else from you. But please, take your time. Thank you!

kayrus commented:
@victorpcamargo thanks for the patience. I reviewed your PR, and I have a lot of concerns.

* There is no "region" argument implemented for both resource and data source
* You're trying to implement shared zones function directly in terraform provider rather than implement them in gophercloud and reuse them
* You're not using the existing `dnsClientSetAuthHeader` function for sudo
* Shared zone resource import test is not implemented

I started a review, but then decided to fix the concerns myself, but at some point it appeared that I started to reimplement everything. Unfortunately I have a bunch of tasks with higher priorities, thus I have to leave my changes as is and switch my focus to other tasks.

Here are my WIP branches for https://github.com/kayrus/gophercloud/tree/list-zone-share
And for provider https://github.com/kayrus/terraform-provider-openstack/tree/feature/shared_zones

Feel free to continue the shared zones implementation and submit a PR to gophercloud.

And again, apologies for late response.

victorpcamargo commented:
@kayrus I had to be away for a while this month and wasn't able to reply to you in time. Sorry about that!
Thanks for all your effort in making this work.
